### ISSUE TITLE
FR: Gesture overview

### ISSUE THREAD

**Issue by LucyOne** [koreader/koreader#7788]
I love KOReader's gesture manager and I have a lot of gestures configured. However, I often forget the ones that I don't use that often. It takes at least 5 taps to go to the gesture manager and look up the configuration.

I would like to have an overview of all configured gestures accessible through the menu on the top level, i.e. with two or three taps, and through a gesture.

At the moment, I'm using a hand-written cheat sheet. An option to save the gesture overview as a printable file would be nice, too.

**Comment by NiLuJe** [comment]
If you like reading lua: `cat /mnt/onboard/.adds/koreader/settings/gestures.lua` ;p.

**Comment by Frenzie** [comment]
I created this overview when I wrote multiswipes btw:
https://github.com/koreader/koreader/wiki/multiswipes

![](https://user-images.githubusercontent.com/202757/54091240-104e0680-437e-11e9-939f-3f8e4d150e31.png)

**Comment by husnoo** [comment]
On a related note, I'm trying to write a plugin, which will have two buttons in the menus, that I want each to be called by a gesture. I can't seem to call any plugin menu items from gestures at all. Is this supported?

Otherwise I'd have to look into adding buttons to the ui to do the thing I'm trying to do.

**Comment by pazos** [comment]
> On a related note, I'm trying to write a plugin, which will have two buttons in the menus, that I want each to be called by a gesture. I can't seem to call any plugin menu items from gestures at all. Is this supported?
> 
> Otherwise I'd have to look into adding buttons to the ui to do the thing I'm trying to do.

Yes. It is supported. Have a look at the calibre plugin, for instance.

In particular https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/main.lua#L63-L67, which register 3 actions on the dispatcher: https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/main.lua#L28-L43

That function is called on plugin's init.

**Comment by poire-z** [comment]
We don't associate directly menu entries to gestures.
We associate events to gestures (technically, we asssociate the event to a generic module, Dispatcher, that will make all its known events into the gesture menu), and we have event handlers that will get the even and do their stuff.
But a menu entry can have as a callback the same event hander.
Some simple example (with the same code made available to menu and dispatcher):
https://github.com/koreader/koreader/blob/bc0a55f093dae3974e0576e866a041bf23597ee2/plugins/systemstat.koplugin/main.lua#L267-L288

**Comment by husnoo** [comment]
Thanks guys! I have just been hacking through calibre, perception and hello plugins. Is this the place to ask for help with these things? Or do you guys have a discord?

**Comment by Frenzie** [comment]
A [Gitter](https://gitter.im/koreader/koreader) (top of README ;-) ).

**Comment by shuvashish76** [comment]
Suggestions:

1. Similar to https://github.com/koreader/koreader/issues/8701#issuecomment-1019272528 i.e. on screen long window to preview your gestures.

![Image](https://github.com/user-attachments/assets/13cf8328-8246-4cda-8385-7bcbcde91fb4)
e.g. https://github.com/shiltemann/cheatsheets/blob/master/awesome-wm.md

I've used [awesomeWM](https://github.com/awesomeWM/awesome) (lua based WM), which has [a default keybinding](https://awesomewm.org/doc/manpages/awesome.1.html#:~:text=Mod4%20+%20s) that shows all your keybindings as cheetsheet which picks changes from their `rc.lua` [config file](https://awesomewm.org/doc/manpages/awesomerc.5.html#:~:text=SYNOPSIS).

2. Gestures as Table format, show to the user as an image/svg file.

For now I've saved them as `.ods` file

<details><summary>Details</summary>
<p>

Gestures | Location | File Browser | Reader
-- | -- | -- | --
Tap corner | Top left | NA | TOC
  | Top right | NA | book map
  | Bottom left | history | history
  | Bottom right | favorites | favorites
  |   |   |  
Long tap on corner (all about statistics only) | Top left | NA | reading statistics : current book
  | Top right | reading statistics : show time range | reading statistics : show time range
  | Bottom left | reading statistics : show today’s timeline | reading statistics : show today’s timeline
  | Bottom right | reading statistics : show calendar view | reading statistics : show calendar view
  |   |   |  
Double Tap (reader specific actions only) | Left Side | NA | turn page -10
  | Right side | NA | turn page +10
  | Top left | NA | bookmarks
  | Top right | NA | toggle bookmarks
  | Bottom left | NA | book status
  | Bottom right | NA | page browser
  |   |   |  
Two-finger tap corner | [All Disabled] |   |  
  |   |   |  
One-finger swipe (all about managing screen and lights only) | Short diagonal swipe | full screen refresh | full screen refresh
  | Left edge down | decrease fontlight brightness: gesture distance | decrease fontlight brightness: gesture distance
  | Left edge up | increase fontlight brightness: gesture distance | increase fontlight brightness: gesture distance
  | Right edge down | increase fontlight warmth:  gesture distance | increase fontlight warmth:  gesture distance
  | Right edge up | decrease fontlight warmth:  gesture distance | decrease fontlight warmth:  gesture distance
  | Top edge right | open next file in last book folder | open next file in last book folder
  | Top edge left | open previous file in last book folder | open previous file in last book folder
  | Bottom edge right | toggle night mode | toggle night mode
  | Bottom edge left | toggle fontlight | toggle fontlight
  |   |   |  
Two-finger swipe | [All Disabled]Besides following 2 |   |  
  | ⬇️ | - sleep | - sleep
  | ⬆️ | - exit sleep screen | - exit sleep screen
  |   |   |  
Spread and pinch | Spread |   | - increase font size in reflowable docs - manual zoom mode in fixed fixed layout docs
  | Pinch |   | - decrease font size in reflowable docs - page full zoom in fixed layout docs
  |   |   |  
Two-finger rotation | Rotate clockwise | rotate by 90 CCW | rotate by 90 CCW
  | Rotate counterclockwise | rotate by 90 CW | rotate by 90 CW
  |   |   |  
Multiswipes | [Defaults] |   |  
  |   |   |  
Custom multiswipes | V | open vocabulary builder | open vocabulary builder
  | Λ | Profile-X | Profile-X
  | < | calibre metadata search | calibre metadata search
  |   |   |  
Settings > Status bar > |   | NA | long press on status bar to skim

Note: You can [set different actions for same gesture for FM and Reader](https://github.com/koreader/koreader/issues/12111#issuecomment-2197007152) but I prefer same action as it's easier to remember.
Note2: For simplicity  I prefer single action for all my gestures but possible to set multiple actions.

</p>
</details>

3. Generate mind-map from the `gestures.lua` file similar to https://markmap.js.org/repl

@offset-torque any better alternative/proposal idea? Blender cheet-sheet looks very complex but indeed it is a complex software.

**Comment by offset-torque** [comment]
We have two problems that needs to be solved here:
- **Action based lookup:** User knows that they assigned an action but can't remember its gesture. "What was the gesture for **Export all notes in current book**?"
- **Gesture based lookup:** User knows that they assigned a gesture but can't remember its action. "What does **double-tap on bottom left corner do**?"

An ideal interface for this is a sortable table like this. In this mockup, users can tap the headers to sort the actions (case: user knows the action wants to know the assigned gesture) or sort the gestures (case: user knows the gesture and wants to know the assigned action). In this image it is sorted alphabetically according to the gesture names for example:

![Image](https://github.com/user-attachments/assets/b5cdaf27-eb15-4da1-94eb-7e13b6c99538)

But our widgets are pretty limited AFAIK and this kind of dialog might be hard or impossible to create.

Luckily this dialog is basically a table. So we can generate a HTML document which contains this table and show that to the user, just like our **"font test document"**. We can add a single menu item like `Taps and gestures > Show assigned gestures` which will create this document on-the-fly and show it to the user. This document will be static so there won't be any sorting, but we can include both versions of the table: Sorted by action and sorted by gesture.

Advantage of this approach is its simplicity. We don't need to change the current menu system much and no need to create/populate/paginate dialogs. And users can copy this document to their computer and see it in their browser. They can even print it out from there for reference (For this use case it would better to create it under the Home folder of library so it would be easy to find).

So technically same as your second suggestion but instead of an image we can show it as a document which is much more versatile, easier to create (we already have the code from the font test generator) and has better UX (no need to zoom-in or drag an image around, just turn the page).

**Comment by shuvashish76** [comment]
> We can add a single menu item like `Taps and gestures > Show assigned gestures` which will create this document on-the-fly and show it to the user.

Yes I thought about this but how quickly it can generate/open the HTML file if the user changing gestures frequently (e.g. when initially testing KOR gestures with lots of changes)?

**Comment by offset-torque** [comment]
Since this would be a fairly simple document to generate and render, my rough guess is around 500ms (generation+opening) on my 5 year old Libra. When factoring in the refresh rate of e-ink starting from 300ms this should feel instant. Of course someone working on the code can give proper estimates but shouldn't be much worse.

For comparison, font test document on my device which is around 200 pages and much more complex with a ToC, generates and opens in ~2 seconds total.

**Comment by Frenzie** [comment]
I'm not quite sure I understand what the concern might be in any case. Regardless how fast or slow it'd be (and I agree it'd be fast), the suggestions above are extremely similar?

**Comment by hius07** [comment]
It can be our usual key-value table. Is it useful?

![Image](https://github.com/user-attachments/assets/c05f7a3c-8d9f-412b-ab2c-9e175788625a)

![Image](https://github.com/user-attachments/assets/12d381ce-a6c7-44f7-ac8a-d9fab7acec61)

![Image](https://github.com/user-attachments/assets/fe58ffe2-fc53-4cbc-8996-b6545cba676f)

**Comment by shuvashish76** [comment]
- Only show FM related gestures overview while on FM and show Reader related gestures overview while on Reader?
- Add additional Action "Gesture overview" under General section?

**Comment by hius07** [comment]
Multiswipes

![Image](https://github.com/user-attachments/assets/9eba48b8-91c2-4d14-8f1c-a77c08cb5ea8)

**Comment by offset-torque** [comment]
Looks pretty useful **hius07**. One improvement would be the removal of empty categories/gestures.
They create a bit of confusion by crowding the list unnecessarily.

![Image](https://github.com/user-attachments/assets/22353586-1c67-4d5e-9e2c-0960e2fefa78)

**Comment by hius07** [comment]
Okay, hide empty sections.
Show both File browser and Reader gestures in one table?
Or separately, depending on the current mode?

**Comment by shuvashish76** [comment]
Personally I don't think all in one table is useful as @offset-torque mentioned the intention of this feature is to quickly lookup some specific gesture/action not memorize them **all**. Besides that (from the screenshots) Reader specific gestures starts from page 2. Which means while on reader view we need one extra tap to view Reader gestures which is unnecessary.

**Comment by shuvashish76** [comment]
Also do consider custom multiswipes as well.

**Comment by offset-torque** [comment]
I just noticed that they are together. I agree with **shuvashish76** that context dependent should be better, since we have the ability now. 

New users will probably launch this dialog from the reader interface while they are setting up other things. It might be confusing to see the other mode's gestures in the first page, especially if they are not aware that our gestures are not global or if they miss the **File browser** title which is not very prominent due to the widget limitations.

**Comment by shuvashish76** [comment]
> [awesomeWM](https://github.com/awesomeWM/awesome) (lua based WM), which has [a default keybinding](https://awesomewm.org/doc/manpages/awesome.1.html#:~:text=Mod4%20+%20s) that shows all your keybindings as cheetsheet which picks changes from their rc.lua [config file](https://awesomewm.org/doc/manpages/awesomerc.5.html#:~:text=SYNOPSIS).
>
> Add additional Action "Gesture overview" under General section?

Any thought on this? Or you think it's unnecessary? @hius07

**Comment by mergen3107** [comment]
Hey guys,

What do you think about adding a popup window with something like "You haven't used KOReader is a while. Would you like to have a look at Gesture Overview?" if device was asleep for 1 week (1 month?).

Context is here (https://github.com/koreader/koreader/issues/14108#issuecomment-3133120618), essentially I just forgot there was a gesture assigned and confused the long delay of opening Text Editor with a delay before a crash.

**Comment by mergen3107** [comment]
> Okay, hide empty sections.
Show both File browser and Reader gestures in one table?
Or separately, depending on the current mode?

I think, the way it works now is good. However, I think there should be hits (for gestures overall), that Reader and File Browser gestures are two separate sets. On a quick thought, here are options:
a) Instead of one common `Gesture overview` have two, `Gesture overview (reader)`  and `Gesture overview (File browser)` - and only one active while the other is grayed out depending on context.
b) In each Gesture Overview window, have an asterisk or a exclamation mark in a circle - something attractive for users to click and find out a text saying "For File Browser (reader) gestures list, open Gesture Overview from there".
c) Keep single common `Gesture overview` but add an asterisk - for users to long press it.

Just my 2¢

**Comment by shuvashish76** [comment]
> a) Instead of one common Gesture overview have two, Gesture overview (reader) and Gesture overview (File browser) - and only one active while the other is grayed out depending on context.

There are separate for Reader and File Browser.

> b) In each Gesture Overview window, have an asterisk or a exclamation mark in a circle - something attractive for users to click and find out a text saying "For File Browser (reader) gestures list, open Gesture Overview from there".

You (the user) opening Gesture Overview option either from menu or via gesture so it's a conscious choice of what you're doing. So I don't think adding such (an asterisk or a exclamation mark in a circle) visual presentation improves anything.
Besides that when you open gesture overview - while on reader - it shows "Gestures in reader", if you open it from FM - it shows "Gestures in file browser" as header.

**Comment by mergen3107** [comment]
> There are separate for Reader and File Browser.

Yes, I (as an advanced user by this point) understand all that.
What I am suggesting is supposed to help a user who sees gestures overview for the first time.
There have been a lot of question in the past when users didn't understand that they need to setup gestures twice: one for reader, one for file browser

**Comment by shuvashish76** [comment]
> What do you think about adding a popup window with something like "You haven't used KOReader is a while. Would you like to have a look at Gesture Overview?" if device was asleep for 1 week (1 month?).

I guess optional reminder only for Gesture overview is okay but then again why not check it from the menu?
Personal opinion on it https://github.com/koreader/koreader/issues/8800#issuecomment-2943614451

> Again it's too much IMO. Better I'd to just check Gesture overview than such reminders troubling me while I'm in flow of reading a book. Reminders can be configurable but still I'd prefer distraction free reading.

Tutorial/Training system discussed in #8800 thread.

**Comment by mergen3107** [comment]
> then again why not check it from the menu?

IMHO, this is fine as long as a user _remembers_ about it :D so this action is active. My suggestion is a passive event.

Anyway, I don't have hard attachments to what I suggested. After all, my point of view at this point might be trying to oversimplify things too much, because I don't even remember "how to be a newcomer to KOReader" anymore T_T

**Comment by shuvashish76** [comment]
> There have been a lot of question in the past when users didn't understand that they need to setup gestures twice: one for reader, one for file browser.

Yes I agree even KOReader devs are confused !? 🫣 #12111
